src/synced_context.cpp

```diff
@@ -200,12 +208,18 @@ void synced_context::set_is_simultaneous()
 	is_simultaneous_ = true;
 }
 
-bool synced_context::can_undo()
+bool synced_context::undo_blocked()
 {
 	// this method should only works in a synced context.
 	assert(is_synced());
-	// if we called the rng or if we sent data of this action over the network already, undoing is impossible.
-	return (!is_simultaneous_) && (randomness::generator->get_random_calls() == 0);
+	// if we called the rng, undoing is blocked.
+	// if we sent data of this action over the network already, undoing is blocked.
+	// if the game has ended, undoing is blocked.
+	// if the turn has ended undoing is blocked.
+	return is_simultaneous_
+	    || (randomness::generator->get_random_calls() == 0)
```

Comment:
Shouldn't this be `!= 0`?

The previous condition was:

`(!is_simultaneous_) && (randomness::generator->get_random_calls() == 0)`

By DeMorgan's law, negating that expression gives:

`(is_simultaneous_) || (randomness::generator->get_random_calls() != 0))`

Reply:
Yes you are right